runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java

```diff
@@ -45,9 +46,12 @@
   private final ListeningExecutorService executor;
   private final ServerConfiguration configuration;
   private final ServerFactory serverFactory;
+  private GrpcFnServer<InMemoryJobService> jobServer;
+  private GrpcFnServer<BeamFileSystemArtifactStagingService> artifactStagingServer;
 
-  private static class ServerConfiguration {
-    @Option(name = "--job-host", required = true, usage = "The job server host string")
+  /** Configuration for the jobServer. */
+  public static class ServerConfiguration {
+    @Option(name = "--job-host", usage = "The job server host string")
```

Comment:
Ports can be chosen dynamically if not provided.
Help facilitate launch of multiple jobServices on the same machine by finding an unused port.

Finding and managing unused ports reliably is surprisingly difficult. The major challenge is that once you pick a port without using it, it can be picked and used by some other process without your knowledge.

Please let me know of any other ways/library to get dynamic unused ports in tests case.

Reply:
Gotcha.  that makes sense.

I have no idea about other ways to manage unused ports (short of writing our own manager), so I think you're right that this is the way to do it.

Reply:
Is this a local port? If so, it's quite easy to do in Java by allocating a new TCP server socket at port 0. You can then query it for the chosen port. This works as long as you are able to pass that socket directly to the service implementation. (Note that you can do this with gRPC and we have a Beam-specific tool that does this for you).

If you're trying to find an unused port for a non-Java server or for a server that will run separately (i.e., for which you do not directly control the socket), you can use the same server allocation trick and note the chosen port number. You can then kill the server socket and pass that port into your new server's configuration. Note that doing this approach introduces a race condition but will work in most cases (especially with retries).

Reply:
That seems to be possible.
Comparing the approach 
1. Optional port and query the port back
2. Passing a socket to the flink driver

I feel option 1 seems to be simpler while keeping initialization flexible.